File: editor/src/main/java/com/libreshockwave/editor/selection/SelectionManager.java
package com.libreshockwave.editor.selection;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Tracks the current selection state in the editor and notifies listeners.
 */
public class SelectionManager {

    private final List<SelectionListener> listeners = new CopyOnWriteArrayList<>();
    private SelectionEvent currentSelection = SelectionEvent.none();

    public void addListener(SelectionListener listener) {
        listeners.add(listener);
    }

    public void removeListener(SelectionListener listener) {
        listeners.remove(listener);
    }

    public SelectionEvent getCurrentSelection() {
        return currentSelection;
    }

    public void select(SelectionEvent event) {
        this.currentSelection = event;
        for (SelectionListener listener : listeners) {
            listener.selectionChanged(event);
        }
    }

    public void clearSelection() {
        select(SelectionEvent.none());
    }
}

File: cpp/src/editor/selection/SelectionManager.cpp
#include "libreshockwave/editor/selection/SelectionManager.hpp"

#include <algorithm>

namespace libreshockwave::editor::selection {

void SelectionManager::addListener(SelectionListener* listener) {
    listeners_.push_back(listener);
}

void SelectionManager::removeListener(SelectionListener* listener) {
    const auto found = std::find(listeners_.begin(), listeners_.end(), listener);
    if (found != listeners_.end()) {
        listeners_.erase(found);
    }
}

const SelectionEvent& SelectionManager::currentSelection() const {
    return currentSelection_;
}

void SelectionManager::select(const SelectionEvent& event) {
    currentSelection_ = event;
    const auto listeners = listeners_;
    for (SelectionListener* listener : listeners) {
        if (listener != nullptr) {
            listener->selectionChanged(event);
        }
    }
}

void SelectionManager::clearSelection() {
    select(SelectionEvent::none());
}

} // namespace libreshockwave::editor::selection
